### ISSUE TITLE
HttpLoggingInterceptor crashes on invalid charsets

### ISSUE THREAD

**Issue by rfc2822** [square/okhttp#1978]
I have encountered a server that returns responses like this:

```
HTTP/1.1 200 OK
Date: Sat, 07 Nov 2015 11:34:54 GMT
Server: Apache
Strict-Transport-Security: max-age=63072000; includeSubdomains; preload
X-Powered-By: PHP/5.4.41
MS-Author-Via: DAV
X-Sabre-Version: 1.8.7
…
Content-Type: text/html; charset=0
```

The charset value causes `HttpLoggingInterceptor` to fail, because it calls `Charset.forName` which throws an `UnsupportedCharsetException` for the value "0". Neither `MediaType.charset()` nor `HttpLoggingInterceptor` handle this exception, so the logging interceptor crashes.

I see two possible methods of handling that:

* `MediaType.charset()` could handle the exception – but which charset should be assumed?
* Maybe that's not so wise. Instead, it should be pointed out that `MediaType.charset()` throws `UnsupportedMediaException`. `HttpLoggingInterceptor` (and maybe other occurrences in the code) should handle this `UnsupportedMediaException` (see PR #1977).

**Comment by JakeWharton** [comment]
It seems unfortunate to crash in here for something like this. We could just fall back to attempting to use UTF-8 in this case.

**Comment by rfc2822** [comment]
I don't think that's appropriate, because the data may be in another charset, and would get misinterpreted and corrupted. Nobody wants corrupted logs with invalid characters. Instead, the `UnsupportedCharsetException` should be documented and handled appropriately.

**Comment by JakeWharton** [comment]
What is "handled appropriately" in terms of a logger then? Displaying
nothing?

**Comment by rfc2822** [comment]
If the charset is not known, special characters (i.e. everything except US-ASCII chars), should be displayed as numeric values, i.e. "text with special chars yb[52][34]wer"

**Comment by JakeWharton** [comment]
I don't really like that as a default behavior. I'd rather just replace it
with the stacktrace at that point.

**Comment by dave-r12** [comment]
I think this one was closed with #2255

**Comment by JakeWharton** [comment]
:+1:
